src/filemanager.cpp

```diff
@@ -0,0 +1,2301 @@
+
+// ---------------------------------------------------------------------------
+
+class DiskChunkCache {
+    PJ_CONTEXT *ctx_ = nullptr;
+    std::string path_{};
+    sqlite3 *hDB_ = nullptr;
+    std::string thisNamePtr_{};
+    std::unique_ptr<SQLite3VFS> vfs_{};
+
+    explicit DiskChunkCache(PJ_CONTEXT *ctx, const std::string &path);
+
+    bool createDBStructure();
+    bool checkConsistency();
+    bool get_links(sqlite3_int64 chunk_id, sqlite3_int64 &link_id,
+                   sqlite3_int64 &prev, sqlite3_int64 &next,
+                   sqlite3_int64 &head, sqlite3_int64 &tail);
+    bool update_links_of_prev_and_next_links(sqlite3_int64 prev,
+                                             sqlite3_int64 next);
+    bool update_linked_chunks(sqlite3_int64 link_id, sqlite3_int64 prev,
+                              sqlite3_int64 next);
+    bool update_linked_chunks_head_tail(sqlite3_int64 head, sqlite3_int64 tail);
+
+    DiskChunkCache(const DiskChunkCache &) = delete;
+    DiskChunkCache &operator=(const DiskChunkCache &) = delete;
+
+  public:
+    static std::unique_ptr<DiskChunkCache> open(PJ_CONTEXT *ctx);
+    ~DiskChunkCache();
+
+    sqlite3 *handle() { return hDB_; }
+    std::unique_ptr<SQLiteStatement> prepare(const char *sql);
+    bool move_to_head(sqlite3_int64 chunk_id);
+    bool move_to_tail(sqlite3_int64 chunk_id);
+    void closeAndUnlink();
+};
+
+// ---------------------------------------------------------------------------
+
+static bool pj_context_get_grid_cache_is_enabled(PJ_CONTEXT *ctx) {
+    pj_load_ini(ctx);
+    return ctx->gridChunkCache.enabled;
+}
+
+// ---------------------------------------------------------------------------
+
+static long long pj_context_get_grid_cache_max_size(PJ_CONTEXT *ctx) {
+    pj_load_ini(ctx);
+    return ctx->gridChunkCache.max_size;
+}
+
+// ---------------------------------------------------------------------------
+
+static int pj_context_get_grid_cache_ttl(PJ_CONTEXT *ctx) {
+    pj_load_ini(ctx);
+    return ctx->gridChunkCache.ttl;
+}
+
+// ---------------------------------------------------------------------------
+
+std::unique_ptr<DiskChunkCache> DiskChunkCache::open(PJ_CONTEXT *ctx) {
+    if (!pj_context_get_grid_cache_is_enabled(ctx)) {
+        return nullptr;
+    }
+    const auto cachePath = pj_context_get_grid_cache_filename(ctx);
+    if (cachePath.empty()) {
+        return nullptr;
+    }
+
+    auto diskCache =
+        std::unique_ptr<DiskChunkCache>(new DiskChunkCache(ctx, cachePath));
+    if (!diskCache->hDB_)
+        diskCache.reset();
+    return diskCache;
+}
+
+// ---------------------------------------------------------------------------
+
+DiskChunkCache::DiskChunkCache(PJ_CONTEXT *ctx, const std::string &path)
+    : ctx_(ctx), path_(path), vfs_(SQLite3VFS::create(true, false, false)) {
+    if (vfs_ == nullptr) {
+        return;
+    }
+    sqlite3_open_v2(path.c_str(), &hDB_,
```

Comment:
It might be better to move all of this into an `initialize()` method instead of having it in the ctor. As it is now, if sqlite3_open_v2 fails to open the db, are we partially constructed? Is ever method of DiskChunkCache supposed to check if hDB_ is good?

Reply:
Moved to initialize(). The constructor being private and used by a static open() method, there is no risk that methods will be used if initialize() fails as it will return a nullptr object